temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowRunTaskHandler.java

```diff
@@ -363,4 +373,36 @@ public void cancel(HistoryEvent cancelEvent) {
       replayWorkflowExecutor.handleWorkflowExecutionCancelRequested(cancelEvent);
     }
   }
+
+  private static class LocalActivityMeteringHelper {
+    private final Map<String, AtomicInteger> firstWftActivities = new HashMap<>();
+    private final Map<String, AtomicInteger> nonFirstWftActivities = new HashMap<>();
+    private final Set<String> completed = new HashSet<>();
+
+    private void newWFTStarting() {
```

Comment:
getNonfirstAttempts and newWFTStarting are not atomic, I think it's better to make them thread-safe, because it's fairly simple here.

To do that, getNonfirstAttempts should decrease the nonFirstWftActivities counters by the counts it read from them.
newWFTStarting should set 0 only in counters that are copied from firstWftActivities.

Reply:
Done :+1: